compiler/src/dmd/parse.d

```diff
@@ -6246,10 +6246,20 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
             }
         case TOK.if_:
             {
+                AST.Statement _init;
                 nextToken();
                 check(TOK.leftParenthesis);
                 auto param = parseAssignCondition();
-                auto condition = parseExpression();
```

Comment:
Silly question: why does the old code do `parseAssignCondition` followed by `parseExpression`? Does `parseAssignCondition` not eat tokens?

Unlike with `for` which (assuming its there) unconditionally parses a statement before the first semicolon, I think here we would need to determine if there is a semicolon present before deciding wether or not to sparse a statement or an expression here.

Reply:
`parseAssignCondition` only eats tokens if it sits on a storage class. If it does not, it just returns `null`. It also does not parse an initializer.

AFAIU, in `if(auto x = e)`, `parseAssignCondition` parses the `auto x =` part, and `parseExpression` parses the `e` part.